From: SGI
Subject: Select MMU_NOTIFIER for SGI_XP
References: bnc#573803

In SLES11 GA, CONFIG_SGI_GRU did the selects MMU_NOTIFIER for us.  This had the
side benefit of enabling us to use our external kernel module XPMEM.  That
kernel module is GPL'd, but not included in upstream kernel.  It has been
submitted to the linux-mm mailing list in the past.  XPMEM is enabling for
SGI's MPT.  We include XPMEM on our ProPack distribution including its source
code.  The request is to get MMU_NOTIFIER enabled again in SLES11 SP1.

Acked-by: Jiri Kosina <jkosina@suse.cz>

Index: linux-2.6.32-SLE11-SP1/drivers/misc/Kconfig
===================================================================
--- linux-2.6.32-SLE11-SP1.orig/drivers/misc/Kconfig
+++ linux-2.6.32-SLE11-SP1/drivers/misc/Kconfig
@@ -166,6 +166,7 @@ config SGI_XP
 	select IA64_UNCACHED_ALLOCATOR if IA64_GENERIC || IA64_SGI_SN2
 	select GENERIC_ALLOCATOR if IA64_GENERIC || IA64_SGI_SN2
 	select SGI_GRU if X86_64 && SMP
+	select MMU_NOTIFIER
 	---help---
 	  An SGI machine can be divided into multiple Single System
 	  Images which act independently of each other and have
