Subject: dasd: move diag kmsg to generic dasd kmsg
References: BNC#563999
From: Gerald Schaefer <geraldsc@de.ibm.com>

Symptom: The dasd-diag messages show up as dasd-eckd or dasd-fba messages.
Problem: The diag discipline does not have a own driver name it uses
             dasd-eckd or dasd-fba as driver.
Solution: Move diag messages to the generic dasd part so that dasd-diag
             messages show up as dasd messages.
Acked-by: John Jolly <jjolly@suse.de>

---
 Documentation/kmsg/s390/dasd      |  123 ++++++++++++++++++++++++++++++++++++++
 Documentation/kmsg/s390/dasd-diag |  118 ------------------------------------
 drivers/s390/block/dasd_diag.c    |   42 ++++++------
 3 files changed, 145 insertions(+), 138 deletions(-)

Index: linux-sles11sp1/Documentation/kmsg/s390/dasd
===================================================================
--- linux-sles11sp1.orig/Documentation/kmsg/s390/dasd
+++ linux-sles11sp1/Documentation/kmsg/s390/dasd
@@ -464,3 +464,126 @@
  * User action:
  * Note the reason code and contact your support organization.
 */
+
+
+------------------------------------------------------------------------------------
+/* dasd_diag */
+
+/*?
+ * Text: "%s: A 64-bit DIAG call failed\n"
+ * Severity: Warning
+ * Parameter:
+ *   @1: bus ID of the DASD
+ * Description:
+ * 64-bit DIAG calls require a 64-bit z/VM version.
+ * User action:
+ * Use z/VM 5.2 or later or set the sysfs 'use_diag' attribute of the DASD to 0
+ * to switch off DIAG.
+ */
+
+/*?
+ * Text: "%s: Accessing the DASD failed because of an incorrect format (rc=%d)\n"
+ * Severity: Warning
+ * Parameter:
+ *   @1: bus ID of the DASD
+ *   @2: return code
+ * Description:
+ * The format of the DASD is not correct.
+ * User action:
+ * Check the device format. For details about the return code see the
+ * section about the INITIALIZE function for DIAGNOSE Code X'250'
+ * in "z/VM CP Programming Services". If you cannot resolve the error, note
+ * the return code and contact your support organization.
+ */
+
+/*?
+ * Text: "%s: New DASD with %ld byte/block, total size %ld KB%s\n"
+ * Severity: Informational
+ * Parameter:
+ *   @1: bus ID of the DASD
+ *   @2: bytes per block
+ *   @3: size
+ *   @4: access mode
+ * Description:
+ * A DASD with the indicated block size and total size has been set online.
+ * If the DASD is configured as read-only to the real or virtual hardware,
+ * the message includes an indication of this hardware access mode. The
+ * hardware access mode is independent from the 'readonly' attribute of
+ * the device in sysfs.
+ * User action:
+ * None.
+ */
+
+/*?
+ * Text: "%s: DIAG ERP failed with rc=%d\n"
+ * Severity: Warning
+ * Parameter:
+ *   @1: bus ID of the DASD
+ *   @2: return code
+ * Description:
+ * An error in the DIAG processing could not be recovered by the error
+ * recovery procedure (ERP) of the DIAG discipline.
+ * User action:
+ * Note the return code, check for related I/O errors, and report this problem
+ * to your support organization.
+ */
+
+/*?
+ * Text: "%s: DIAG initialization failed with rc=%d\n"
+ * Severity: Warning
+ * Parameter:
+ *   @1: bus ID of the DASD
+ *   @2: return code
+ * Description:
+ * Initializing the DASD with the DIAG discipline failed. Possible reasons for
+ * this problem are that the device has a device type other than FBA or ECKD,
+ * or has a block size other than one of the supported sizes:
+ * 512 byte, 1024 byte, 2048 byte, or 4096 byte.
+ * User action:
+ * Ensure that the device can be written to and has a supported device type
+ * and block size. For details about the return code see the section about
+ * the INITIALIZE function for DIAGNOSE Code X'250' in "z/VM CP Programming
+ * Services". If you cannot resolve the error, note the error code and contact
+ * your support organization.
+ */
+
+/*?
+ * Text: "%s: Device type %d is not supported in DIAG mode\n"
+ * Severity: Warning
+ * Parameter:
+ *   @1: bus ID of the DASD
+ *   @2: device type
+ * Description:
+ * Only DASD of type FBA and ECKD are supported in DIAG mode.
+ * User action:
+ * Set the sysfs 'use_diag' attribute of the DASD to 0 and try again to access
+ * the DASD.
+ */
+
+/*?
+ * Text: "Discipline %s cannot be used without z/VM\n"
+ * Severity: Informational
+ * Parameter:
+ *   @1: discipline name
+ * Description:
+ * The discipline that is specified with the dasd= kernel or module parameter
+ * is only available for Linux instances that run as guest operating
+ * systems of the z/VM hypervisor.
+ * User action:
+ * Remove the unsupported discipline from the parameter string.
+ */
+
+/*?
+ * Text: "%s: The access mode of a DIAG device changed to read-only\n"
+ * Severity: Warning
+ * Parameter:
+ *   @1: bus ID of the DASD
+ * Description:
+ * A device changed its access mode from writeable to
+ * read-only while in use.
+ * User action:
+ * Set the device offline, ensure that the device is configured correctly in
+ * z/VM, then set the device online again.
+ */
+
+------------------------------------------------------------------------------------
Index: linux-sles11sp1/Documentation/kmsg/s390/dasd-diag
===================================================================
--- linux-sles11sp1.orig/Documentation/kmsg/s390/dasd-diag
+++ /dev/null
@@ -1,118 +0,0 @@
-/* dasd_diag */
-
-/*?
- * Text: "%s: A 64-bit DIAG call failed\n"
- * Severity: Warning
- * Parameter:
- *   @1: bus ID of the DASD
- * Description:
- * 64-bit DIAG calls require a 64-bit z/VM version.
- * User action:
- * Use z/VM 5.2 or later or set the sysfs 'use_diag' attribute of the DASD to 0
- * to switch off DIAG.
- */
-
-/*?
- * Text: "%s: Accessing the DASD failed because of an incorrect format (rc=%d)\n"
- * Severity: Warning
- * Parameter:
- *   @1: bus ID of the DASD
- *   @2: return code
- * Description:
- * The format of the DASD is not correct.
- * User action:
- * Check the device format. For details about the return code see the
- * section about the INITIALIZE function for DIAGNOSE Code X'250'
- * in "z/VM CP Programming Services". If you cannot resolve the error, note
- * the return code and contact your support organization.
- */
-
-/*?
- * Text: "%s: New DASD with %ld byte/block, total size %ld KB%s\n"
- * Severity: Informational
- * Parameter:
- *   @1: bus ID of the DASD
- *   @2: bytes per block
- *   @3: size
- *   @4: access mode
- * Description:
- * A DASD with the indicated block size and total size has been set online.
- * If the DASD is configured as read-only to the real or virtual hardware,
- * the message includes an indication of this hardware access mode. The
- * hardware access mode is independent from the 'readonly' attribute of
- * the device in sysfs.
- * User action:
- * None.
- */
-
-/*?
- * Text: "%s: DIAG ERP failed with rc=%d\n"
- * Severity: Warning
- * Parameter:
- *   @1: bus ID of the DASD
- *   @2: return code
- * Description:
- * An error in the DIAG processing could not be recovered by the error
- * recovery procedure (ERP) of the DIAG discipline.
- * User action:
- * Note the return code, check for related I/O errors, and report this problem
- * to your support organization.
- */
-
-/*?
- * Text: "%s: DIAG initialization failed with rc=%d\n"
- * Severity: Warning
- * Parameter:
- *   @1: bus ID of the DASD
- *   @2: return code
- * Description:
- * Initializing the DASD with the DIAG discipline failed. Possible reasons for
- * this problem are that the device has a device type other than FBA or ECKD,
- * or has a block size other than one of the supported sizes:
- * 512 byte, 1024 byte, 2048 byte, or 4096 byte.
- * User action:
- * Ensure that the device can be written to and has a supported device type
- * and block size. For details about the return code see the section about
- * the INITIALIZE function for DIAGNOSE Code X'250' in "z/VM CP Programming
- * Services". If you cannot resolve the error, note the error code and contact
- * your support organization.
- */
-
-/*?
- * Text: "%s: Device type %d is not supported in DIAG mode\n"
- * Severity: Warning
- * Parameter:
- *   @1: bus ID of the DASD
- *   @2: device type
- * Description:
- * Only DASD of type FBA and ECKD are supported in DIAG mode.
- * User action:
- * Set the sysfs 'use_diag' attribute of the DASD to 0 and try again to access
- * the DASD.
- */
-
-/*?
- * Text: "Discipline %s cannot be used without z/VM\n"
- * Severity: Informational
- * Parameter:
- *   @1: discipline name
- * Description:
- * The discipline that is specified with the dasd= kernel or module parameter
- * is only available for Linux instances that run as guest operating
- * systems of the z/VM hypervisor.
- * User action:
- * Remove the unsupported discipline from the parameter string.
- */
-
-/*?
- * Text: "%s: The access mode of a DIAG device changed to read-only"
- * Severity: Warning
- * Parameter:
- *   @1: bus ID of the DASD
- * Description:
- * A device changed its access mode from writeable to
- * read-only while in use.
- * User action:
- * Set the device offline, ensure that the device is configured correctly in
- * z/VM, then set the device online again.
- */
Index: linux-sles11sp1/drivers/s390/block/dasd_diag.c
===================================================================
--- linux-sles11sp1.orig/drivers/s390/block/dasd_diag.c
+++ linux-sles11sp1/drivers/s390/block/dasd_diag.c
@@ -8,7 +8,7 @@
  *
  */
 
-#define KMSG_COMPONENT "dasd-diag"
+#define KMSG_COMPONENT "dasd"
 
 #include <linux/stddef.h>
 #include <linux/kernel.h>
@@ -147,16 +147,16 @@ dasd_diag_erp(struct dasd_device *device
 	rc = mdsk_init_io(device, device->block->bp_block, 0, NULL);
 	if (rc == 4) {
 		if (!(device->features & DASD_FEATURE_READONLY)) {
-			dev_warn(&device->cdev->dev,
-				 "The access mode of a DIAG device changed"
-				 " to read-only");
+			pr_warning("%s: The access mode of a DIAG device "
+				   "changed to read-only\n",
+				   dev_name(&device->cdev->dev));
 			device->features |= DASD_FEATURE_READONLY;
 		}
 		rc = 0;
 	}
 	if (rc)
-		dev_warn(&device->cdev->dev, "DIAG ERP failed with "
-			    "rc=%d\n", rc);
+		pr_warning("%s: DIAG ERP failed with "
+			    "rc=%d\n", dev_name(&device->cdev->dev), rc);
 }
 
 /* Start a given request at the device. Return zero on success, non-zero
@@ -372,8 +372,9 @@ dasd_diag_check_device(struct dasd_devic
 		private->pt_block = 2;
 		break;
 	default:
-		dev_warn(&device->cdev->dev, "Device type %d is not supported "
-			    "in DIAG mode\n", private->rdc_data.vdev_class);
+		pr_warning("%s: Device type %d is not supported "
+			   "in DIAG mode\n", dev_name(&device->cdev->dev),
+			   private->rdc_data.vdev_class);
 		rc = -EOPNOTSUPP;
 		goto out;
 	}
@@ -414,8 +415,8 @@ dasd_diag_check_device(struct dasd_devic
 		private->iob.flaga = DASD_DIAG_FLAGA_DEFAULT;
 		rc = dia250(&private->iob, RW_BIO);
 		if (rc == 3) {
-			dev_warn(&device->cdev->dev,
-				"A 64-bit DIAG call failed\n");
+			pr_warning("%s: A 64-bit DIAG call failed\n",
+				   dev_name(&device->cdev->dev));
 			rc = -EOPNOTSUPP;
 			goto out_label;
 		}
@@ -424,8 +425,9 @@ dasd_diag_check_device(struct dasd_devic
 			break;
 	}
 	if (bsize > PAGE_SIZE) {
-		dev_warn(&device->cdev->dev, "Accessing the DASD failed because"
-			 " of an incorrect format (rc=%d)\n", rc);
+		pr_warning("%s: Accessing the DASD failed because of an "
+			   "incorrect format (rc=%d)\n",
+			   dev_name(&device->cdev->dev), rc);
 		rc = -EIO;
 		goto out_label;
 	}
@@ -443,18 +445,18 @@ dasd_diag_check_device(struct dasd_devic
 		block->s2b_shift++;
 	rc = mdsk_init_io(device, block->bp_block, 0, NULL);
 	if (rc && (rc != 4)) {
-		dev_warn(&device->cdev->dev, "DIAG initialization "
-			"failed with rc=%d\n", rc);
+		pr_warning("%s: DIAG initialization failed with rc=%d\n",
+			   dev_name(&device->cdev->dev), rc);
 		rc = -EIO;
 	} else {
 		if (rc == 4)
 			device->features |= DASD_FEATURE_READONLY;
-		dev_info(&device->cdev->dev,
-			 "New DASD with %ld byte/block, total size %ld KB%s\n",
-			 (unsigned long) block->bp_block,
-			 (unsigned long) (block->blocks <<
-					  block->s2b_shift) >> 1,
-			 (rc == 4) ? ", read-only device" : "");
+		pr_info("%s: New DASD with %ld byte/block, total size %ld "
+			"KB%s\n", dev_name(&device->cdev->dev),
+			(unsigned long) block->bp_block,
+			(unsigned long) (block->blocks <<
+					 block->s2b_shift) >> 1,
+			(rc == 4) ? ", read-only device" : "");
 		rc = 0;
 	}
 out_label:
