Subject: dasd: support DIAG access for read-only devices
References: BNC#563999,LTC#57147
From: Gerald Schaefer <geraldsc@de.ibm.com>

Symptom: If a DASD is attached as read-only under z/VM and it is
             configured to be used with the DIAG discipline, then setting
             the device online will fail. The message log will show a
             message 'DIAG initialization failed (rc=4)'.
Problem: The DIAG return code 4 indicates that the initialization was
             successful and the device is read-only, but the DASD device
             driver interprets this as an error.
Solution: Fix the DIAG device initialization to accept a return value
             of 4 as indication of success.
Acked-by: John Jolly <jjolly@suse.de>

---

 drivers/s390/block/dasd_diag.c |   19 ++++++++++++++++---
 1 file changed, 16 insertions(+), 3 deletions(-)

Index: linux-sles11sp1/drivers/s390/block/dasd_diag.c
===================================================================
--- linux-sles11sp1.orig/drivers/s390/block/dasd_diag.c
+++ linux-sles11sp1/drivers/s390/block/dasd_diag.c
@@ -145,6 +145,15 @@ dasd_diag_erp(struct dasd_device *device
 
 	mdsk_term_io(device);
 	rc = mdsk_init_io(device, device->block->bp_block, 0, NULL);
+	if (rc == 4) {
+		if (!(device->features & DASD_FEATURE_READONLY)) {
+			dev_warn(&device->cdev->dev,
+				 "The access mode of a DIAG device changed"
+				 " to read-only");
+			device->features |= DASD_FEATURE_READONLY;
+		}
+		rc = 0;
+	}
 	if (rc)
 		dev_warn(&device->cdev->dev, "DIAG ERP failed with "
 			    "rc=%d\n", rc);
@@ -433,16 +442,20 @@ dasd_diag_check_device(struct dasd_devic
 	for (sb = 512; sb < bsize; sb = sb << 1)
 		block->s2b_shift++;
 	rc = mdsk_init_io(device, block->bp_block, 0, NULL);
-	if (rc) {
+	if (rc && (rc != 4)) {
 		dev_warn(&device->cdev->dev, "DIAG initialization "
 			"failed with rc=%d\n", rc);
 		rc = -EIO;
 	} else {
+		if (rc == 4)
+			device->features |= DASD_FEATURE_READONLY;
 		dev_info(&device->cdev->dev,
-			 "New DASD with %ld byte/block, total size %ld KB\n",
+			 "New DASD with %ld byte/block, total size %ld KB%s\n",
 			 (unsigned long) block->bp_block,
 			 (unsigned long) (block->blocks <<
-					  block->s2b_shift) >> 1);
+					  block->s2b_shift) >> 1,
+			 (rc == 4) ? ", read-only device" : "");
+		rc = 0;
 	}
 out_label:
 	free_page((long) label);
