Subject: dasd: enable prefix independent of pav support
References: BNC#565612
From: Gerald Schaefer <geraldsc@de.ibm.com>

Symptom: High Performance Ficon is not usable when PAV is not enabled
Problem: The prefix command is only enabled with enabled PAV support.
Solution: The prefix command is always enabled.
Acked-by: John Jolly <jjolly@suse.de>

---

 drivers/s390/block/dasd_eckd.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/drivers/s390/block/dasd_eckd.c
+++ b/drivers/s390/block/dasd_eckd.c
@@ -1013,9 +1013,9 @@ static struct dasd_ccw_req *dasd_eckd_bu
 	}
 	psf_ssc_data = (struct dasd_psf_ssc_data *)cqr->data;
 	psf_ssc_data->order = PSF_ORDER_SSC;
-	psf_ssc_data->suborder = 0x40;
+	psf_ssc_data->suborder = 0xc0;
 	if (enable_pav) {
-		psf_ssc_data->suborder |= 0x88;
+		psf_ssc_data->suborder |= 0x08;
 		psf_ssc_data->reserved[0] = 0x88;
 	}
 	ccw = cqr->cpaddr;
