From: Jiri Kosina <jkosina@suse.cz>
Subject: CFQ: turn 'low_latency' knob off by default
References: bnc#572834

Turn 'low_latency' knob off by default for CFQ IO scheduler for
performance reasons.

More discussion about this being more 'desktop'-oriented setting
can be found on

  http://groups.google.com/group/linux.kernel/browse_thread/thread/b3cc422601e9bf58/0d590d5bf8337540?lnk=raot

Acked-by: Jan Kara <jack@suse.cz>
Signed-off-by: Jiri Kosina <jkosina@suse.cz>

Index: linux-2.6.32-SLE11-SP1/block/cfq-iosched.c
===================================================================
--- linux-2.6.32-SLE11-SP1.orig/block/cfq-iosched.c
+++ linux-2.6.32-SLE11-SP1/block/cfq-iosched.c
@@ -2543,7 +2543,7 @@ static void *cfq_init_queue(struct reque
 	cfqd->cfq_slice[1] = cfq_slice_sync;
 	cfqd->cfq_slice_async_rq = cfq_slice_async_rq;
 	cfqd->cfq_slice_idle = cfq_slice_idle;
-	cfqd->cfq_latency = 1;
+	cfqd->cfq_latency = 0;
 	cfqd->hw_tag = 1;
 	cfqd->last_end_sync_rq = jiffies;
 	return cfqd;
