From: Greg Kroah-Hartman <gregkh@suse.de>
Subject: kabi: acpi: guard acpi_power_register changes

The structure is now using the reserved field so it names it properly.
This has no affect on the kabi other than the fields changing names so
protect it from the kabi checker.

Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>
---
 include/acpi/processor.h |    4 ++++
 1 file changed, 4 insertions(+)

--- a/include/acpi/processor.h
+++ b/include/acpi/processor.h
@@ -48,7 +48,11 @@ struct acpi_power_register {
 	u8 space_id;
 	u8 bit_width;
 	u8 bit_offset;
+#ifdef __GENKSYMS__
+	u8 reserved;
+#else
 	u8 access_size;
+#endif
 	u64 address;
 } __attribute__ ((packed));
 
