From: Dmitry Torokhov <dtor@vmware.com>
Date: Fri, 23 Apr 2010 13:18:08 -0400
Subject: [PATCH] VMware Balloon driver
References: bnc#669740, fate#311976
Git-commit: 7b6c7c51549d18bd14b9c712fc98b42fefd0670f

This is a standalone version of VMware Balloon driver.  Ballooning is a
technique that allows hypervisor dynamically limit the amount of memory
available to the guest (with guest cooperation).  In the overcommit
scenario, when hypervisor set detects that it needs to shuffle some
memory, it instructs the driver to allocate certain number of pages, and
the underlying memory gets returned to the hypervisor.  Later hypervisor
may return memory to the guest by reattaching memory to the pageframes and
instructing the driver to "deflate" balloon.

The driver has been shipping for a number of years and users running on
VMware platform will have it installed as part of VMware Tools even if it
will not come from a distribution, thus there should not be additional
risk in pulling the driver into mainline.  The driver will only activate
if host is VMware so everyone else should not be affected at all.

The driver has been present in mainline since 2.6.34.

Signed-off-by: Dmitry Torokhov <dtor@vmware.com>
Acked-by: Michal Hocko <mhocko@suse.cz>

Automatically created from "patches.drivers/VMware-Balloon-driver-2.6.32.28.patch" by xen-port-patches.py

 drivers/misc/Kconfig |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- sle11sp2-2011-05-11.orig/drivers/misc/Kconfig	2011-05-11 13:11:16.000000000 +0200
+++ sle11sp2-2011-05-11/drivers/misc/Kconfig	2011-05-11 13:28:25.000000000 +0200
@@ -249,7 +249,7 @@ config EP93XX_PWM
 
 config VMWARE_BALLOON
 	tristate "VMware Balloon Driver"
-	depends on X86
+	depends on X86 && !XEN
 	help
 	  This is VMware physical memory management driver which acts
 	  like a "balloon" that can be inflated to reclaim physical pages
